docs: properties for invariant testing (#12)
defi-wonderland/optimism: packages/contracts-bedrock/test/invariants/echidna/superc20/PROPERTIES.md

supERC20 properties
===================

Valid state
-----------

| id    | description                                                           | halmos | echidna |
| ----- | -----                                                                 | -----  | -----   |

> |   0   | calls to sendERC20 succeed as long as caller has enough balance       |        |         |

Reviewer: ```suggestion
|   0   | calls to `sendERC20` succeed as long as caller has enough balance       |        |         |
```

Variable transition
-------------------

| id    | description                                                           | halmos | echidna |
| ----- | -----                                                                 | -----  | -----   |

> |   1   | sendERC20 decreases the token's totalSupply in the source chain       |        |         |
> |   2   | relayERC20 increases the token's totalSupply in the destination chain |        |         |
> |   3   | only calls to `convert()` can increase the total supply across chains |        |         |

Reviewer: ```suggestion
|   1   | `sendERC20` decreases the token's total supply in the source chain       |        |         |
|   2   | `relayERC20` increases the token's total supply in the destination chain |        |         |
|   3   | only calls to `convert()` and `mint()` can increase the total supply across chains |        |         |
```

High level
----------

| id    | description                                                                                                  | halmos | echidna |
| ----- | -----                                                                                                        | -----  | -----   |

> |   4   | sum of total supply across all chains is always `<=` to `convert()`ed amount                                 |        |         |

Reviewer: ```suggestion
|   4   | sum of total supply across all chains is always `<=` to `convert()`-ed and `mint()`-ed amount                                 |        |         |
```

Author: I'm not sure I follow you on this one, what `mint()` are you referring to?

with this property I'm trying to describe the high-level behaviour of the whole system, where the only place where tokens can be minted should be inside `convert`

I guess we could also define a separate 'Valid State' property local to the supERC20 where we check total supply is the sum of all mints, if we don't get that for free by including  ToB ERC20 properties

Reviewer: I'm referring to the `OptimismSuperchainERC20#mint` function, that is only callable by the standard bridge on the L1 -> L2 flow.
Should we take that into consideration when defining the total amount of the token invariant?

|   5   | tokens `sendERC20`-ed to a chain can be `relayERC20`-ed as long as the source chain is in the dependency set |        |         |

Unit test
---------

| id    | description                                                                | halmos | echidna |
| ----- | -----                                                                      | -----  | -----   |

> |   6   | supERC20 token address does not depend on chainID                          |        |         |

Reviewer: The invariant wouldn't be something like
```suggestion
|   6   | supERC20 token addresses are the same in all the different chains |        |         |
```
?

Author: I believe this is currently clearer to map to a halmos proof, like so:

```
check_addressDoesNotDependOnChainId(remote, name, symbol, decimals, chainId1, chainID2){
  vm.assume(chainId1 != chainId2);
  assertEq(factory.deploy(remote, name, symbol, chainId1),  factory.deploy(remote, name, symbol, chainId2));
}
```

wdyt? it certainly _feels_ math-ier

Reviewer: ah perfect, if that's simpler for halmos let's keep like that
